Initial commit

nwithan8 opened the pull request:
This is the initial work for `EasyVCR`'s Java version, designed as similar as [`easyvcr-csharp`](https://github.com/EasyPost/easyvcr-csharp) as possible. Includes:
- Advanced settings:
  - Censor sensitive headers, query and body parameters in requests and responses
  - Custom match rules when determining if a request matches a recording
  - Custom or original delay when replaying a request
- Direct client access, or built-in VCR utility to track cassettes and settings
- Set VCR mode via environmental variable

Currently only supports [Java 8's `Http(s)UrlConnection`](https://docs.oracle.com/javase/8/docs/api/java/net/HttpURLConnection.html).

Future support planned for [Java 11's `HttpClient`](https://www.baeldung.com/java-9-http-client) and [Apache's `HttpClient`](https://hc.apache.org/httpcomponents-client-4.5.x/quickstart.html)

Justintime50 commented:
Needs lint fixed, then will check.

Justintime50 commented on src/main/java/com/easypost/easyvcr/internalutilities/Files.java:
```diff
@@ -0,0 +1,83 @@
+            myWriter.write(string);
+            myWriter.close();
+        } catch (IOException ignored) {
+            System.out.println("An error occurred while writing to the file.");
```
This should raise an error and not print. Printing is rarely what we want in an app.

nwithan8 replied:
Good catch

Justintime50 commented on src/main/java/com/easypost/easyvcr/requestelements/HttpVersion.java:
```diff
@@ -0,0 +1,65 @@
+     *
+     * @param version the HTTP ProtocolVersion
+     */
+    public HttpVersion(ProtocolVersion version) {
```
What is this used for?

nwithan8 replied:
`HttpURLConnection` stores its HTTP version info in a class called `ProtocolVersion`. This is to take that `ProtocolVersion` and convert it to our EasyVCR HttpVersion class.

Justintime50 commented on src/main/java/com/easypost/easyvcr/requestelements/Response.java:
```diff
@@ -0,0 +1,434 @@
+
+            @Override
+            public void setParams(HttpParams httpParams) {
+                // not implemented
```
For all these `not implemented` functions, do we want to throw a `not implemented` error instead?

nwithan8 replied:
We can't throw any error, since that would change the function signature, which we can't do because it's an override.

nwithan8 replied:
I was able to implement a few more, but for the most part, the user only received this object if it's a replayed request/response, and at this point in the workflow, the request was already made, so a user modifying these attributes after the fact doesn't make much sense in my mind anyway.

Justintime50 commented on src/test/java/HttpUrlConnectionTest.java:
```diff
@@ -0,0 +1,213 @@
+    }
+
+    @Test
+    public void testDelay() throws Exception {
```
I'd hope we'll quickly revisit these tests and break them out. There's way too much happening in a single test here.

nwithan8 replied:
It's the exact same tests as C# :)

nwithan8 replied:
Ultimately they are, there's just a lot of setup steps in each test, but few asserts

nwithan8 commented on .github/workflows/ci.yml:
```diff
@@ -0,0 +1,33 @@
+          distribution: "zulu"
+          java-version: ${{ matrix.javaversion }}
+      - name: Build and test with Maven
+        run: mvn --batch-mode install -Dgpg.skip=true -Dcheckstyle.skip=true
```
It's skipping the checkstyle checks, which are run in the separate linter step

nwithan8 commented on src/main/java/com/easypost/easyvcr/Cassette.java:
```diff
@@ -0,0 +1,175 @@
+     * Overwrite an existing interaction on this cassette, or add a new one if it doesn't exist.
+     * @param httpInteraction The interaction to write to the cassette
+     * @param matchRules The rules to match the interaction against
+     * @param bypassSearch If true, the cassette will not be searched for an existing interaction
```
It's an internal shortcut, if we know an matching recording doesn't exist already (i.e we already attempted replaying, and now we're for sure in record mode), it skips the part attempting to find a match. This is not user facing, but helps makes processing faster.

nwithan8 commented on src/main/java/com/easypost/easyvcr/MatchRules.java:
```diff
@@ -0,0 +1,235 @@
+    }
+
+    /**
+     * Default rule is to match on the method and URL.
```
The default for C# is also just url and method, a hold-over from Scotch.
Processing the body introduces a lot more overhead during the match process.

Justintime50 replied:
I'd highly encourage we follow other VCR solutions and match on bodies by default as well.

nwithan8 replied:
Actually: https://github.com/vcr/vcr/blob/4d14c470442c2037f0fa544f384ab10ddd7e3781/lib/vcr/request_matcher_registry.rb#L9

Justintime50 replied:
Good find. We must be explicitly adding the body matcher elsewhere. Ok, this default is fine. Thanks for the info.

nwithan8 commented on src/main/java/com/easypost/easyvcr/clients/httpurlconnection/RecordableHttpURLConnection.java:
```diff
@@ -0,0 +1,1401 @@
+     */
+    public RecordableHttpURLConnection(URL url, Proxy proxy, Cassette cassette, Mode mode,
+                                       AdvancedSettings advancedSettings) throws IOException {
+        // this super is not used
```
Because it's needed since this class extends the standard HttpURLConnection

nwithan8 commented on src/main/java/com/easypost/easyvcr/clients/httpurlconnection/RecordableHttpURLConnection.java:
```diff
@@ -0,0 +1,1401 @@
+        // not in cassette, go to real connection
+        return this.connection.getOutputStream();
+    }
+}
```
`RecordableHttpURLConnection` is a wrapper around `HttpURLConnection` that, as a result, needs to extend all the functions of `HttpURLConnection`. Those that don't need to be recorded/replayed simply get passed down to the wrapper `HttpURLConnection` object.

nwithan8 commented on src/main/java/com/easypost/easyvcr/clients/httpurlconnection/RecordableURL.java:
```diff
@@ -0,0 +1,300 @@
+/**
+ * A recordable URL, wrapper for RecordableHttpURLConnection and RecordableHttpsURLConnection.
+ */
+public final class RecordableURL {
```
I'm not sure your question

Justintime50 replied:
You inadvertently answered it elsewhere

nwithan8 commented on pom.xml:
```diff
@@ -0,0 +1,259 @@
+            <version>RELEASE</version>
+            <scope>compile</scope>
+        </dependency>
+        <!-- https://mvnrepository.com/artifact/commons-beanutils/commons-beanutils -->
```
Just an informative comment link to the artifact on Maven

nwithan8 commented on src/main/java/com/easypost/easyvcr/interactionconverters/HttpUrlConnectionInteractionConverter.java:
```diff
@@ -0,0 +1,117 @@
+            return new ResponseAndTime(response, milliseconds);
+        } catch (URISyntaxException | IOException ignored) {
+        }
+        // FIXME: make it so this won't blow up on error
```
Ultimately parses the same way ([an IDE finds it the same way it finds a TODO](https://docs.microsoft.com/en-us/visualstudio/mac/task-comments?view=vsmac-2019)), but okay.

Justintime50 replied:
It's dangerous to assume the the magic features of any one IDE is how everyone will access or use something. EasyPost is 99% terminal based editors internally (our team is the exception) so although it may work in yours, it doesn't work in others. Certainly if I go on a `TODO fix spree`, I'll miss anything that's titled `FIXME` because I'm only searching for TODO. We also use `TODO` almost exclusively internally for that reason.

Justintime50 commented on src/main/java/com/easypost/easyvcr/Mode.java:
```diff
@@ -0,0 +1,5 @@
+package com.easypost.easyvcr;
+
+public enum Mode {
+    Bypass, Auto, Record, Replay
+}
```
No luck on this?

Justintime50 commented on src/main/java/com/easypost/easyvcr/clients/httpurlconnection/RecordableHttpURLConnection.java:
```diff
@@ -0,0 +1,1401 @@
+        }
+    }
+
+    /*
+    private static String getParamsString(Map<String, String> params) {
+        StringBuilder result = new StringBuilder();
+
+        for (Map.Entry<String, String> entry : params.entrySet()) {
+            result.append(URLEncoder.encode(entry.getKey(), StandardCharsets.UTF_8));
+            result.append("=");
+            result.append(URLEncoder.encode(entry.getValue(), StandardCharsets.UTF_8));
+            result.append("&");
+        }
+
+        String resultString = result.toString();
+        return resultString.length() > 0 ? resultString.substring(0, resultString.length() - 1) : resultString;
+    }*/
```
This wasn't addressed.

Justintime50 commented on src/main/java/com/easypost/easyvcr/internalutilities/Files.java:
```diff
@@ -0,0 +1,83 @@
+     */
+    public static void createFileIfNotExists(String filePath) throws IOException {
+        try {
+
```
This wasn't addressed

Justintime50 replied:
This was not addressed

jchen293 commented on src/main/java/com/easypost/easyvcr/interactionconverters/HttpUrlConnectionInteractionConverter.java:
```diff
@@ -0,0 +1,117 @@
+            request.setMethod(method);
+            request.setUri(new URI(uriString));
+            request.setHeaders(headers);
+            // TODO: Capture and censor body
+            // request.setBody(body);
```
Remove the comment and uncomment tine 47

nwithan8 replied:
This damn line opened a whole can of worms.

It's been dealt with.

jchen293 commented on src/main/java/com/easypost/easyvcr/interactionconverters/HttpUrlConnectionInteractionConverter.java:
```diff
@@ -0,0 +1,117 @@
+            return request;
+        } catch (URISyntaxException ignored) {
+        }
+        // FIXME: make it so this won't blow up on error
```
Should we remove it?

jchen293 commented on src/main/java/com/easypost/easyvcr/internalutilities/Utils.java:
```diff
@@ -0,0 +1,117 @@
+        headerKeys.add("upgrade");
+        headers.addAll(headerKeys);
+
+        // String v = getNetProperty("jdk.httpclient.allowRestrictedHeaders");
```
This need to be removed

Justintime50 commented:
Drafting this until it's ready.

Justintime50 reviewed (changes requested):
Code seems fine, noticed a few things:

1. We need to ignore the `target` directory so we don't accidentally include that dir in code
2. Can we save the cassettes as pretty JSON instead of a hard to read blob?
3. The README needs at least cursory info (how to build, a single example for now, how to test, and what the thing is) - we can do the rest as we release
4. There are a couple of build warnings and a couple dozen IDE warnings that we should fix (I'd be fine fixing these in a separate PR for all our upcoming initial release items, just wanted to call it out). I'd like to avoid these ever creeping in in the first place and always resolving them as we come across them so they don't get reintroduced
5. The tests that actually match bodies are doing GET requests heh, we need to post some data with a body somewhere to ensure the actual body matches instead of naively matching two empty strings. (I did however confirm that if the cassette body and test mismatch it fails and then put it back and it passed so body matching appears to work, we just need to test/guard against it better)

nwithan8 commented:
> Code seems fine, noticed a few things:
> 
> 1. We need to ignore the `target` directory so we don't accidentally include that dir in code
> 2. Can we save the cassettes as pretty JSON instead of a hard to read blob?
> 3. The README needs at least cursory info (how to build, a single example for now, how to test, and what the thing is) - we can do the rest as we release
> 4. There are a couple of build warnings and a couple dozen IDE warnings that we should fix (I'd be fine fixing these in a separate PR for all our upcoming initial release items, just wanted to call it out). I'd like to avoid these ever creeping in in the first place and always resolving them as we come across them so they don't get reintroduced
> 5. The tests that actually match bodies are doing GET requests heh, we need to post some data with a body somewhere to ensure the actual body matches instead of naively matching two empty strings. (I did however confirm that if the cassette body and test mismatch it fails and then put it back and it passed so body matching appears to work, we just need to test/guard against it better)

1. Yes.
2. Yes.
3. Yes.
4. Took a brief look, addressed some.
5. Fixed.

Justintime50 commented on README.md:
```diff
@@ -1 +1,248 @@
+#### Credit
+
+- [C# EasyVCR](https://github.com/EasyPost/easyvcr-csharp), upon which this library is based
+- [Scotch by Martin Leech](https://github.com/mleech/scotch), whose core functionality inspired the C# version of EasyVCR
```
Nit: this line isn't necessary to call out but either way
